### PR TITLE
Add sshbarrier to ssh plugin

### PR DESCRIPTION
**Pull request by wangdian**
If sshbarrier is defined, the ssh will wait until all configured taskrole containers can be attached.

### CODE REVIEW

File: src/kube-runtime/src/plugins/ssh/README.md [hunk 1]
```diff
@@ -10,9 +10,12 @@ extras:
...
       parameters:
         jobssh: boolean
+        sshbarrier:
+          - taskrole
         userssh: 
           type: string
           value: string
 ```
 - jobssh: true to enable job container wise ssh, false to disable.
+- sshbarrier: wait until can ssh to all defined taskrole's job containers. Only valid when jobssh is true.
```
Comment (fanyangCS) [inline src/kube-runtime/src/plugins/ssh/README.md]: please define sshbarrier for the whole job level, instead of the task role level.

Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/README.md]: Now sshbarrier is a boolean to define if barrier is enabled.
Use optional param sshbarriertaskroles to identify specific taskroles that barrier watches to meet advanced requirements.

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+  instancePair=(${i//:/ })
+  taskRole=${instancePair[0]}
+  index=${instancePair[1]}
+
+  if [[ $taskRole = $FC_TASKROLE_NAME ]] && [[ $index = $FC_TASK_INDEX ]]; then
+    continue
+  fi
+
+# If barrier task roles defined, then only check instances for defined task roles. Otherwise check all instances.
+  if [ ${#barrierTaskRoles[@]} != 0 ]; then
```
Comment (Binyang2014) [inline src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Use `if [[  ]]` here?

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+    if [[ " ${barrierTaskRoles[@]} " =~ " ${value} " ]]; then
```
Comment (Binyang2014) [inline src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Why there is a space before and after the var?

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+      instanceToCheck+=("${taskRole}-${index}")
+    fi
+  else
+    instanceToCheck+=("${taskRole}-${index}")
+  fi
+done
+
+retryCount=0
+while [ ${#instanceToCheck[@]} -ne 0 ]
```
Comment (Binyang2014) [inline src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: use `double bracket` here?

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+  instanceFailed=()
+  for instance in "${instanceToCheck[@]}"; do
+    check_ssh_connection "$instance"
+    if [ $? -ne 0 ]; then
+      instanceFailed+=("$instance")
+    fi
+  done
+  instanceToCheck=(${instanceFailed[*]}) 
+
+  ((retryCount++))
```
Comment (Binyang2014) [inline src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Can break the loop if instanceToCheck is empty?

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+    instanceToCheck+=("${taskRole}-${index}")
+  fi
+done
+
+retryCount=0
+while [ ${#instanceToCheck[@]} -ne 0 ]
+do
+  if [ $retryCount -ge $MAX_RETRY_COUNT ]; then
+    echo "SSH barrier reaches max retry count. Failed instances: ${instanceToCheck[*]} Exit..."
+    exit 240
```
Comment (Binyang2014) [inline src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Please make sure such log redirect to the runtime error log, or print to stderr

Just thinking about using specific exit code range for runtime plugin. For sshbarrier, we can give the code range [10, 20). And for each exit code, we need to udpate the error spec like this:
```yaml
- code: 240
  phrase: xxxx
  issuer: PAI_RUNTIME
  causer:PAI_RUNTIME
  type: PLATFORM_FAILURE
  stage: xxxx
  behavior: PERMANENT
  reaction: NEVER_RETRY
  reason: "SSH barrier reaches max retry count"
  repro:
  - "xxx"
  solution:
  - "xxx"
  pattern:
    runtimeContainerPatterns:
    - exitCode: 10
      platformLogRegex: "SSH barrier reaches max retry count"
```

Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Added

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh (hunk 2 above)
Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Refine all if

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh (hunk 3 above)
Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: refined

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh (hunk 4 above)
Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Unified code style to double brackets

File: src/kube-runtime/src/plugins/ssh/sshbarrier.sh (hunk 5 above)
Comment (wangdian) [reply src/kube-runtime/src/plugins/ssh/sshbarrier.sh]: Refined

File: src/k8s-job-exit-spec/config/k8s-job-exit-spec.yaml [hunk 7]
```diff
@@ -827,3 +827,21 @@ spec:
   - "PAI Runtime exits with exitcode 1"
   solution:
   - "Contact PAI Dev to fix PAI Runtime bugs"
+
+- code: 240
```
Comment (Binyang2014) [inline src/k8s-job-exit-spec/config/k8s-job-exit-spec.yaml]: Move this to line 739?

Comment (wangdian) [reply src/k8s-job-exit-spec/config/k8s-job-exit-spec.yaml]: Done

**Pull request merged**